src/data/array/defs.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2018 Keeley Hoek. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Keeley Hoek
+-/
+
+/-!
+# Copy functions for arrays
+
+This file provides functions to copy (part) of an array to another array.
+We include monadic variants.
+-/
+
+universe variables u v z
+
+namespace array
+
+variables {k : Type v → Type z} [monad k] {α : Type u} {β : Type v} {n m : ℕ}
+
+/-- Auxiliary function for monadically copying
+part of an array to another array while applying a function.
+If `k` is a monad, `x` and `y` are arrays of lengths `n` and `m`, and `r` is an index,
+then `mmap_copy_aux f r x y` will take the slice `x[r .. (min n m)]`,
+apply the monadic function `f` to it, and write it to the corresponding slice of `y`.
+The result is the modified array `y`, wrapped in the monad `k`. -/
+def mmap_copy_aux (f : α → k β) : ℕ → array n α → array m β → k (array m β)
+| r x y := do if h : r < n ∧ r < m then do
+                let fn : fin n := ⟨r, and.elim_left h⟩,
+                let fm : fin m := ⟨r, and.elim_right h⟩,
+                y ← y.write fm <$> f (x.read fn),
+                have wf : m - (r + 1) < m - r,
+                  from nat.lt_of_succ_le $ by rw [← nat.succ_sub h.2, nat.succ_sub_succ],
+                mmap_copy_aux (r + 1) x y
+              else return y
+using_well_founded {rel_tac := λ _ _, `[exact ⟨_, measure_wf (λ a, m - a.1)⟩]}
+
+/-- Function for monadically copying part of an array to another array while applying a function.
+If `k` is a monad, and `x` and `y` are arrays of lengths `n` and `m`,
+then `array.mmap_copy f x y` will take the first `(min n m)` coefficients of `x`,
+apply the monadic function `f` to them, and write them to the corresponding of `y`.
+The result is the modified array `y`, wrapped in the monad `k`. -/
+def mmap_copy (x : array n α) (y : array m β) (f : α → k β) : k (array m β) :=
+mmap_copy_aux f 0 x y
+
+/-- Function for copying part of an array to another array while applying a function.
+If `x` is an array of length `n`, and `y` is an array of length `m`,
+then `array.map_copy f x y` will take the first `(min n m)` coefficients of `x`, apply `f` to them,
+and write them to the corresponding of `y`.
+
+See `array.copy` for the version with `f = id`. -/
+def map_copy (x : array n α) (y : array m β) (f : α → β) : array m β :=
+@mmap_copy id _ _ _ _ _ x y f
```

Comment:
In the original PR, this function was duplicated for performance reasons.  If you implement it using the generic monadic version, then the VM needs to execute the bind operation etc. of the identity monad in every iteration.

Reply:
Aha... this isn't compiled away in the case of the identity monad? I will reduplicate the function.

Reply:
As far as the VM is concerned a `monad` is just a septuple of functions including `pure`, `bind`, `(>>)` etc (well, actually it is a pair containing a quintuple containing a pair).  The `mmap_copy` takes such a septuple as an argument, and then retrieves `pure` and `map` from this septuple and calls these function in every iteration.

Sometimes the VM can get rid of this indirection by inlining the function (then it sees that we want to compute `@has_pure.pure _ ↑(monad.mk ... id.map ...)` and simplifies this to `id.map`).  Alas, Lean 3 can't inline recursive functions.  Inlining also doesn't help in every case, you don't want to inline thousands of lines of code to get rid of type classes.

Another strategy is called specialization.  With specialization, the compiler creates a copy of the function where type class instance is instantiated with `id.monad`.  (Automating the process that you would now do manually.)  This will be supported in Lean 4 as far as I know.